### PR TITLE
[CALCITE-3621] Push down sort to DB, SQL of Sort rel contains explicit field name instead of * when sort is not top rel

### PR DESCRIPTION
**Pull request by helloppx**
[https://issues.apache.org/jira/browse/CALCITE-3621#](url)
Push down sort to DB, SQL of Sort rel contains explicit field name instead of *
:D

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java [hunk 1]
```diff
@@ -634,6 +634,13 @@ public Result visit(Sort e) {
     }
     Result x = visitChild(0, e.getInput());
     Builder builder = x.builder(e, Clause.ORDER_BY);
+    if (builder.select.getSelectList() == null) {
```
Comment (jinxing64) [inline core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java]: Why this change is necessary ?

File: core/src/main/java/org/apache/calcite/adapter/jdbc/JdbcRules.java [hunk 2]
```diff
@@ -765,10 +766,11 @@ public RelNode convert(RelNode rel) {
      */
     public RelNode convert(Sort sort, boolean convertInputTraits) {
       final RelTraitSet traitSet = sort.getTraitSet().replace(out);
+      RelTraitSet inputTraitSet = traitSet.replace(RelCollations.EMPTY);
```
Comment (jinxing64) [inline core/src/main/java/org/apache/calcite/adapter/jdbc/JdbcRules.java]: Why it's not as below
```
final RelTraitSet inputTraitSet = sort.getInput().getTraitSet().replace(out);
```

Comment (helloppx) [reply core/src/main/java/org/apache/calcite/adapter/jdbc/JdbcRules.java]: Yes, your change is better :D

File: core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java (hunk 1 above)
Comment (helloppx) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java]: When JdbcSort is used, a regression issue will be trigged: JdbcTest#testSelfJoinDifferentColumns()

RelToSqlConverter visit this Sort and then produce a SQL "SELECT * from (t JOIN t0 ...) order by ...)". However t and t0 include one same name field "last_name"

Before SQL:
select e1."full_name"
  from "foodmart"."employee" as e1
  join "foodmart"."employee" as e2 on e1."first_name" = e2."last_name"
order by e1."last_name" limit 3

REL:
JdbcToEnumerableConverter
  JdbcProject(full_name=[$1], last_name=[$3])
    JdbcSort(sort0=[$3], dir0=[ASC], fetch=[3])
      JdbcJoin(condition=[=($2, $0)], joinType=[inner])
        JdbcProject(last_name=[$3])
          JdbcTableScan(table=[[foodmart, employee]])
        JdbcProject(full_name=[$1], first_name=[$2], last_name=[$3])
          JdbcTableScan(table=[[foodmart, employee]])

After SQL:
SELECT "t2"."full_name", "t2"."last_name0" AS "last_name"
FROM (SELECT *
FROM (SELECT "last_name"
FROM "foodmart"."employee") AS "t"
INNER JOIN (SELECT "full_name", "first_name", "last_name"
FROM "foodmart"."employee") AS "t0" ON "t"."last_name" = "t0"."first_name"
ORDER BY "t0"."last_name" NULLS LAST
LIMIT 3) AS "t2"

Comment (jinxing64) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java]: Without the  change in RelToSqlConverter.java, I still passed  the unchanged `testMatchRecognizePatternExpression12`

Comment (helloppx) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java]: This change is for `JdbcTest#testSelfJoinDifferentColumns()` mainly, `testMatchRecognizePatternExpression12` is affected by the side effects, so I have to change it

Comment (jinxing64) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java]: Thanks for explanation ~
Since it's very common to execute sql like `select * from XXX order by col`, shall we wrap the change within a check like below:
```
    if (stack.getLast().r != e) {
      ...
    }
```
Thus we can save the change in`RelToSqlConverterTest.java` and keep the converted Sql String simplier.

Comment (helloppx) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java]: Thank you so much for your help ~ :D

**Comment by helloppx** [comment]
By the way, when we switch "calcite.enable.enumerable" to "false" without this change, this case will be failed. 
`java.sql.SQLException: Error while executing SQL "explain plan for select ename 
from scott.emp 
order by empno": There are not enough rules to produce a node with desired properties: convention=ENUMERABLE, sort=[1].
Missing conversion is JdbcTableScan[sort: [] -> [0]]
There is 1 empty subset: rel#34:Subset#0.JDBC.SCOTT.[0], the relevant part of the original plan is as follows
0:JdbcTableScan(table=[[SCOTT, EMP]])`

File: core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java (hunk 1 above)
Comment (danny0405) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java]: Just curious, shouldn't we generate `SELECT "t2"."full_name", "t2"."last_name0" AS "last_name"` instead of `select *` ?

Why we need a decision `if (stack.getLast().r != e)` it looks weird.

Also, the generated sql outputs one more column than the original, i think it is because we generate the sql using the RelRoot.rel directly. We better fix this.

File: core/src/main/java/org/apache/calcite/adapter/jdbc/JdbcRules.java [hunk 3]
```diff
@@ -765,10 +765,11 @@ public RelNode convert(RelNode rel) {
      */
     public RelNode convert(Sort sort, boolean convertInputTraits) {
       final RelTraitSet traitSet = sort.getTraitSet().replace(out);
+      final RelTraitSet inputTraitSet = sort.getInput().getTraitSet().replace(out);
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/adapter/jdbc/JdbcRules.java]: Move this line below `if (convertInputTraits)` where it is actually used.

File: core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java (hunk 1 above)
Comment (helloppx) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java]: I have updated the description of regission issue again, please help review :D

`if (stack.getLast().r != e)` can make a decision whether currently Sort rel is the outermost rel.
If it is, the select list of generated SQL can be "select *", It is more simplier.
If not, the select list of generated SQL shoud contain explicit column name.

`CalcitePrepareImpl#implement(RelRoot root)` generate SQL by root.rel，but the correctly select field be controled by `root.fields`.

File: core/src/main/java/org/apache/calcite/adapter/jdbc/JdbcRules.java (hunk 3 above)
Comment (helloppx) [reply core/src/main/java/org/apache/calcite/adapter/jdbc/JdbcRules.java]: Thank you, I will update it.

File: core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java (hunk 1 above)
Comment (jinxing64) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java]: Hi Danny ~
I gave the suggestion to keep `select *` when `Sort` is the outmost rel, which I think is much simpler.
I'm ok if you think that Calcite should expand the select list for `Sort`.

Comment (danny0405) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java]: Thanks @jinxing64 , after some research i agree that maybe fixing the sort project list is the easiest way to go, but i would suggest to use `stack.size() == 1` to decide if this sort is a root node instead of `stack.getLast().r != e`.

Comment (helloppx) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java]: Thank you guys, I have updated it.

**Comment by helloppx** [comment]
> The commit message is a little confusing. IMO, it is better to use the title of the jira, namely `JDBC adapter can't push down sort to DB` as commit message.

Thank you @chunweilei , I have updated it.
